### PR TITLE
B#src 5453 fix reset bug

### PR DESCRIPTION
**Pull request by giusebar**
General refactoring, fix bug with reset, fix calibration boresight bug

### CODE REVIEW

File: src/polhemus_ros_driver/polhemus.cpp [hunk 1]
```diff
@@ -142,82 +142,72 @@ int Polhemus::send_saved_calibration(void)
...
+    float calibrated_pitch;
+    std::string calibrated_pitch_param_name = "/calibration/" + name + "_calibration/rotations/station_" + std::to_string(station_id) + "/calibrated_pitch";
+    nh->getParam(calibrated_pitch_param_name, calibrated_pitch);
+
+    float calibrated_yaw;
+    std::string calibrated_yaw_param_name = "/calibration/" + name + "_calibration/rotations/station_" + std::to_string(station_id) + "/calibrated_yaw";
+    nh->getParam(calibrated_yaw_param_name, calibrated_yaw);
+
+    // retrieve current sensor angle
+    tf2::Quaternion current_quaternion;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:169]: can we name it more explanatory? also maybe remove current? to me it suggests there is some previous.. like station_quaternion? not adamant about this one tho

File: src/polhemus_ros_driver/polhemus.cpp [hunk 2]
```diff
@@ -142,82 +142,72 @@ int Polhemus::send_saved_calibration(void)
...
+    {
+      ROS_ERROR_STREAM("Station " << station_id << ": " << msg);
+      return -1;
+      break;
+    }
+
+    double current_roll, current_pitch, current_yaw;
+    tf2::Matrix3x3(current_quaternion).getRPY(current_roll, current_pitch, current_yaw);
+
+    current_roll = (current_roll * 180) / PI;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:184]: station_roll?

File: src/polhemus_ros_driver/polhemus.cpp [hunk 3]
```diff
@@ -142,82 +142,72 @@ int Polhemus::send_saved_calibration(void)
...
+    current_pitch = (current_pitch * 180) / PI;
+    current_yaw = (current_yaw * 180) / PI;
+
+    // compute correction needed to calibrate sensor to 0
+    float correction_roll = current_roll - calibrated_roll;
+    float correction_pitch = current_pitch - calibrated_pitch;
+    float correction_yaw = current_yaw - calibrated_yaw;
+
+    if (name == "viper")
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:193]: bleh

File: src/polhemus_ros_driver/polhemus.cpp [hunk 4]
```diff
@@ -142,82 +142,72 @@ int Polhemus::send_saved_calibration(void)
...
+      define_data_type(DATA_TYPE_QUAT);
     }
     else
     {
-      // no previous calibration exists
-      ROS_WARN("[POLHEMUS] Station could not be found in calibration, please calibrate before proceeding!!!");
+      retval = set_boresight(false, station_id + 1, correction_yaw, correction_pitch, correction_roll);
+    }
+
+    if (retval == RETURN_ERROR)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:204]: reverse order please, its c++

File: src/polhemus_ros_driver/polhemus.cpp [hunk 5]
```diff
@@ -142,82 +142,72 @@ int Polhemus::send_saved_calibration(void)
...
+    {
+      ROS_ERROR("[POLHEMUS] Error sending calibration from file.");
+      return -1;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:207]: no need to break if u return

File: src/polhemus_ros_driver/polhemus.cpp [hunk 6]
```diff
@@ -234,65 +224,53 @@ bool Polhemus::calibrate(std::string boresight_calibration_file)
 
   device_reset();
 
-  if (name == "viper")
+  for (int station_id = 0; station_id < station_count; ++station_id)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:227]: this is almost all duplicated code. Should we wrap in a method?

File: src/polhemus_ros_driver/polhemus.cpp [hunk 7]
```diff
@@ -234,65 +224,53 @@ bool Polhemus::calibrate(std::string boresight_calibration_file)
...
+  
+    std::string calibrated_yaw_param_name = "/calibration/" + name + "_calibration/rotations/station_" + std::to_string(station_id) + "/calibrated_yaw";
+    nh->setParam(calibrated_yaw_param_name, current_yaw);
   }
 
-
   std::string cmd("rosparam dump ");
   cmd += boresight_calibration_file + " /calibration";
   
   if (system(cmd.c_str()))
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:255]: does it return true when it fails? sup with dat

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 8]
```diff
@@ -271,6 +269,18 @@ int discover_vip_pid(libusb_device_handle **usbhnd, vp_usbdevinfo &usbinfo, uint
   return retval;
 }
 
+void reset_and_release_polhemus_device(Polhemus *device, libusb_device_handle *g_usbhnd, vp_usbdevinfo g_usbinfo)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:272]: should be two methods

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 9]
```diff
@@ -304,8 +316,8 @@ int main(int argc, char** argv) {
     retval = discover_vip_pid(&g_usbhnd, g_usbinfo, VENDOR, product_id);
     if (retval == RETURN_ERROR)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:317]: reverse order

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 10]
```diff
@@ -285,6 +295,8 @@ int main(int argc, char** argv) {
   Polhemus *device;
   int retval = RETURN_ERROR;
 
+  bool ready_to_start = true;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:298]: do we need this? shoudnt we justr immediately terminate if its not ready?

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 11]
```diff
@@ -347,13 +358,7 @@ int main(int argc, char** argv) {
   if (retval == RETURN_ERROR)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:358]: reverse

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 12]
```diff
@@ -347,13 +358,7 @@ int main(int argc, char** argv) {
...
   {
     ROS_ERROR("[POLHEMUS] Error resetting device.");
-  }
-
-  retval = device->reset_boresight();
-  if (retval == RETURN_ERROR)
-  {
-    ROS_ERROR("[POLHEMUS] Error resetting boresight.");
-    return 1;
+    return 0;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:361]: all returns should be -1, 0 code means success

Comment (mikramarc) [reply src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:361]: or some other code, just not zero

File: src/polhemus_ros_driver/viper.cpp [hunk 13]
```diff
@@ -279,12 +279,19 @@ int Viper::reset_boresight(void)
 
 tf2::Quaternion Viper::get_quaternion(int index)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/viper.cpp:280]: get quaternion of what? index of what?

File: src/polhemus_ros_driver/viper.cpp [hunk 14]
```diff
@@ -279,12 +279,19 @@ int Viper::reset_boresight(void)
...
+    tf2::Quaternion q(
+        pno.SensFrame(index)->pno.ori[1],
+        pno.SensFrame(index)->pno.ori[2],
+        pno.SensFrame(index)->pno.ori[3],
+        pno.SensFrame(index)->pno.ori[0]);
+    return q;
+  }
+  else
+  {
+    throw "Could not retrieve current sensor orientation, empty sensor frame!";
```
Comment (mikramarc) [inline src/polhemus_ros_driver/viper.cpp:293]: please throw known exception, not just message

**Review by mikramarc — CHANGES_REQUESTED** [review]

File: src/polhemus_ros_driver/polhemus.cpp (hunk 3 above)
Comment (giusebar) [reply src/polhemus_ros_driver/polhemus.cpp:193]: lol

File: src/polhemus_ros_driver/polhemus.cpp (hunk 6 above)
Comment (giusebar) [reply src/polhemus_ros_driver/polhemus.cpp:227]: mm not really, it's quite different, in one we set and in one we get the params. mainly the loop is similar, but there are quite a lot of different things going on

File: src/polhemus_ros_driver/polhemus.cpp (hunk 7 above)
Comment (giusebar) [reply src/polhemus_ros_driver/polhemus.cpp:255]: ye, weird ? !

File: src/polhemus_ros_driver/polhemus.cpp [hunk 15]
```diff
@@ -142,81 +144,69 @@ int Polhemus::send_saved_calibration(void)
...
+    // retrieve current sensor orientation
+    tf2::Quaternion station_quaternion;
+    try
+    {
+      station_quaternion = get_station_quaternion(station_id);
+    }
+    catch(std::runtime_error &error)
+    {
+      ROS_ERROR_STREAM("Caught runtime error for station " << station_id << ": " << error.what());
+      throw error;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:179]: Just throw; is enough i think

File: src/polhemus_ros_driver/polhemus.cpp [hunk 16]
```diff
@@ -142,81 +144,69 @@ int Polhemus::send_saved_calibration(void)
...
-          }
-
-          if (retval == RETURN_ERROR)
-          {
-            ROS_ERROR("[POLHEMUS] Error sending calibration from file.");
-            break;
-          }
-        }
-      }
+      ROS_WARN("[POLHEMUS] No previous calibration data available, please calibrate before proceeding!!!");
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:151]: wow, three exclamation marks, very serious

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp [hunk 17]
```diff
@@ -271,6 +269,12 @@ int discover_vip_pid(libusb_device_handle **usbhnd, vp_usbdevinfo &usbinfo, uint
   return retval;
 }
 
+void release_polhemus_device(Polhemus *device, libusb_device_handle *g_usbhnd, vp_usbdevinfo g_usbinfo)
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:272]: why have this in one method? Just do release device with only device arg, and release usb separately?

File: src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp (hunk 9 above)
Comment (mikramarc) [reply src/polhemus_ros_driver/polhemus_tf_broadcaster.cpp:309]: DO IT

**Review by mikramarc — CHANGES_REQUESTED** [review]

File: src/polhemus_ros_driver/polhemus.cpp [hunk 18]
```diff
@@ -128,11 +128,21 @@ int Polhemus::send_saved_calibration(void)
...
+    ros::Time start_time = ros::Time::now();
 
-    device_reset();
+    while (retval < SENSORS_PER_GLOVE)
+    {
+      retval = receive_pno_data_frame();
+      retval = 4;
+      if (ros::Time::now().toSec() - start_time.toSec() >= CALIBRATE_TIMEOUT_IN_SECS)
+      {
+        ROS_ERROR("[POLHEMUS] Caliration - error getting complete frame in required time.");
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:140]:
```suggestion
        ROS_ERROR("[POLHEMUS] Calibration - error getting complete frame in required time.");
```

File: src/polhemus_ros_driver/polhemus.cpp [hunk 19]
```diff
@@ -128,11 +128,21 @@ int Polhemus::send_saved_calibration(void)
...
   {
-    reset_boresight();
     retval = receive_pno_data_frame();
+    ros::Time start_time = ros::Time::now();
 
-    device_reset();
+    while (retval < SENSORS_PER_GLOVE)
+    {
+      retval = receive_pno_data_frame();
+      retval = 4;
```
Comment (mikramarc) [inline src/polhemus_ros_driver/polhemus.cpp:137]: great code

**Review by mikramarc — CHANGES_REQUESTED** [review]

**Review by mikramarc — APPROVED** [review]

**Review by ToivoS — APPROVED** [review]

**Pull request merged**